cloudify/utils.py

```diff
@@ -914,3 +917,74 @@ def target_plugin_prefix(name, tenant_name, version=None, deployment_id=None):
     # declared always take precedence
     parts += [name, version or '0.0.0']
     return os.path.join(*parts)
+
+
+def parse_utc_datetime_from_expression(time_expression, timezone):
+    if not time_expression:
+        return None
+
+    # time expression is time delta from now
+    if time_expression.startswith('+'):
+        deltas = re.findall(r"(\+\d+\ ?[a-z]+\ ?)", time_expression)
+        if not deltas:
+            raise NonRecoverableError(
+                "{} is not a legal time delta".format(time_expression))
+        date_time = datetime.utcnow().replace(second=0, microsecond=0)
+        for delta in deltas:
+            date_time = parse_and_apply_timedelta(
+                delta.rstrip().lstrip('+'), date_time)
+        return date_time
+
+    # time expression is datetime
+    date_time = parse_schedule_datetime_string(time_expression)
+    if timezone:
+        if timezone not in pytz.all_timezones:
+            raise NonRecoverableError(
+                "{} is not a recognized timezone".format(timezone))
+        return pytz.timezone(timezone).localize(date_time).astimezone(
+            pytz.utc).replace(tzinfo=None)
+    else:
+        ts = time.time()
+        return date_time - (datetime.fromtimestamp(ts) -
+                            datetime.utcfromtimestamp(ts))
+
+
+def parse_timedelta(expr):
+    match = r"^(\d+)\ ?(min(ute)?|h(our)?|d(ay)?|w(eek)?|mo(nth)?|y(ear)?)s?$"
+    parsed = re.findall(match, expr)
+    if not parsed or len(parsed[0]) < 2:
+        raise NonRecoverableError("{} is not a legal time delta".format(expr))
+    return int(parsed[0][0]), parsed[0][1]
+
+
+def parse_and_apply_timedelta(expr, date_time):
```

Comment:
"and" is very bad for functions, but thankfully there's already a separate "parse_timedelta" function that could do the parsing
then all the year/month/week/day/hour could go live in there

Reply:
well, the problem here is that datetime.timedelta doesn't actually handle months and years, so this requires that kind of hybrid function